corgidrp/detector.py

```diff
@@ -520,7 +525,161 @@ def calc_sat_fwc(emgain_arr,fwcpp_arr,fwcem_arr,sat_thresh):
+        full_residuals[ymin:ymax,xmin:xmax][bool_iniszero] = nrr_copy[bool_iniszero]
+        full_residuals[ymin:ymax,xmin:xmax][bool_outisnotzero] = full_residuals[ymin:ymax,xmin:xmax][bool_outisnotzero]
+        
+        
+        
+        full_qe_resel = ndimage.convolve(full_qe, mask)
+        full_residuals_resel = ndimage.convolve(full_residuals,mask)
+        std_resel = (((full_qe_resel - full_residuals_resel)/full_qe_resel))
+    
+    full_residuals[full_residuals==0] = None
+    nx = np.arange(0,std_resel.shape[1])
+    ny = np.arange(0,std_resel.shape[0])
+    nxx,nyy = np.meshgrid(ny,nx)
+    nrr = np.sqrt((nxx-n/2)**2 + (nyy-n/2)**2)
+    full_residuals[nrr>n_pix]= 1
+    std_resel[nrr>n_pix] = 0
+    full_qe[nrr>n_pix]= 1
+    full_qe=np.pad(full_qe, ((n_pad,n_pad),(n_pad,n_pad)), mode='constant',constant_values=(1))
+    full_residuals=np.pad(full_residuals, ((n_pad,n_pad),(n_pad,n_pad)), mode='constant',constant_values=(1))
+    std_resel=np.pad(std_resel, ((n_pad,n_pad),(n_pad,n_pad)), mode='constant',constant_values=(0))
+    
+    return (full_qe,full_residuals,std_resel)
+    
+    
+def create_onsky_flatfield(dataset, planet=None,band=None,up_radius=None,im_size=None,rad_mask1=None, rad_mask4=None,  planet_rad=None, n_pix=None, n_pad=None,N=None):
+    """Turn this dataset of raster scanned image frames of neptune or uranus into on-sky calibrated flat field 
+    
+        Args:
+            dataset (corgidrp.data.Dataset): a dataset of Image frames (L2a-level)
+            planet (str): neptune or uranus
+            band (str): 1 or 4
+            up_radius (int): Number of pixels on either side of centroided planet images (=55 pixels for Neptune and uranus)
+            im_size (int): x-dimension of the planet image (in pixels= 420 for the HST images)
+            N (int): Number of images to be combined for creating a matched filter (defaults to 3, can be changed)
+            rad_mask1 (float): radius in pixels used for creating a mask for band1
+            rad_mask4 (float): radius in pixels used for creating a mask for band4
+            planet_rad (int): radius of the planet in pixels (planet_rad=50 for neptune, planet_rad=65)
+            n_pix (int): Number of pixels in radius covering the Roman CGI imaging FOV defaults to 165 pixels
+            n_pad (int): Number of pixels padded with '1s'  to generate the image size 1024X1024 pixels around imaging FOV (defaults to 302 pixels)
+
+
+    	Returns:
+    		data.FlatField (corgidrp.data.FlatField): a master flat for flat calibration using on sky images of planet in band specified
+    		
+	"""
+    smooth_images=[];raster_images_cent=[];cent=[]; act_cents=[];frames=[]; files=[];
+    for i in range(len(dataset)):
+        target=dataset[i].pri_hdr['TARGET']
+        filter=dataset[i].pri_hdr['FILTER']
+        if planet==target and band==filter: 
```

Comment:
Two things here: 

1. You should expect that all files passed in are indeed for that filter and planet.
2. The behavior of `planet` and `band` should be that if the user doesn't pass in either (i.e., they are None), then we look in the header to determine it. Otherwise, use the planet and band specified by user input. We don't need to compare the input planet/band against what's in the header. This just gives us the flexibility to override the header values, but otherwise, we shouldn't be needing to user planet/band.

Reply:
When I create a `dataset` to be passed to the `create_flatfield` function, it has all the  `.*fits` images from the folder, which might contain both Uranus and Neptune and band 1 and 4. I thought what if the user doesn't have that information ?. The other option is, I do this comparison in `test function` and pass only the dataset that corresponds to the planet and band.

Reply:
I think we should always know which files are the ones we want for a given planet/filter. You can do this in the test function if that simplifies your code.